### PR TITLE
persist_source: Ensure drops also stop subscribe

### LINKED ISSUE #15831: storage, computed: Fast create and drop materialized view leaves task running in computed
### What version of Materialize are you using?

v0.30.0-dev

### How did you install Materialize?

Built from source

### What is the issue?

Executing a fast sequence of create + drop materialized view leaves a tokio task running in computed.

```
> CREATE TABLE t1 (f1 int);
> CREATE MATERIALIZED VIEW v1 AS SELECT MIN(f1) FROM t1; DROP MATERIALIZED VIEW v1;
```
will lead computed to forever try to fetch a snapshot that will never be available:

```
2022-11-01T14:43:16.646303Z  INFO mz_persist_client::internal::machine: snapshot sf2bd7d0d-4bff-469e-b956-6892afee97a3 as of Antichain { elements: [1667313447000] } not yet available for upper Antichain { elements: [1667313447000] } retrying in 16s
```
and this message will repeat every 16 seconds.

In this case `sf2...` is actually the shard of the materialized view, not of `t1`, when computed starts writing to the shard, it first fetches a snapshot of it, which fails in this case and enters an endless retry loop.

See also this [slack thread](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1667314342022869)

It might also be a bug in the envd<->computed interaction, could also be Team Compute.

### PR DESCRIPTION
**Pull request by lluki**
Blocked by #15868. [Slack thread](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1667473791576749)

When a collection is dropped while we are trying to install a persist_sink, it is possible that it hangs forever in `subscribe` because the upper never advances. This PR fixes the problem by using select! to terminate the subscribe.

Fixes #15831

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - None

### CODE REVIEW

**Comment by lluki** [comment]
Unfortunately depending on where the `subscribe().await` is cancelled, computed panics with: 


`
thread 'tokio-runtime-worker' panicked at 'LeasedBatchPart cannot be dropped with lease intact: LeasedBatchPart { shard_id: ShardId(0cb9713c-3f34-476f-aaae-a869c9e43401), reader_id: ReaderId(88a311c8-3881-4708-8337-eaf90ecb7b96), metadata: Snapshot { as_of: [[134, 240, 29, 61, 132, 1, 0, 0]] }, desc: Description { lower: Antichain { elements: [1667472683142] }, upper: Antichain { elements: [1667472683191] }, since: Antichain { elements: [0] } }, key: PartialBatchKey("wcc0e6893-6a30-4756-97f7-cc13819ea40c/pd90b1397-233a-4195-9304-a2fc8eb0a0fa"), leased_seqno: Some(SeqNo(13)) }', /home/lukas/mat/materialize/src/persist-client/src/fetch.rs:392:9
`

**Comment by lluki** [comment]
Stack trace of the crash:
```
environmentd: 2022-11-03T12:29:52.614401Z  INFO mz_adapter::catalog: drop materialized view materialize.public.v1 (u802)
thread 'tokio-runtime-worker' panicked at 'LeasedBatchPart cannot be dropped with lease intact: LeasedBatchPart { shard_id: ShardId(95ca5eb6-bd14-4c5d-906a-83a19be513ae), reader_id: ReaderId(57c18ec7-a43c-4ec2-b529-432f439426e6), metadata: Snapshot { as_of: [[5, 28, 120, 61, 132, 1, 0, 0]] }, desc: Description { lower: Antichain { elements: [1667478592517] }, upper: Antichain { elements: [1667478592585] }, since: Antichain { elements: [0] } }, key: PartialBatchKey("w447a4d0f-f0c2-4759-b687-92b80e7ff036/p73819379-dbf9-4998-929b-0421c15d485f"), leased_seqno: Some(SeqNo(13)) }', /home/lukas/mat/materialize/src/persist-client/src/fetch.rs:392:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
   2: <mz_persist_client::fetch::LeasedBatchPart<T> as core::ops::drop::Drop>::drop
             at ./src/persist-client/src/fetch.rs:392:9
   3: core::ptr::drop_in_place<mz_persist_client::fetch::LeasedBatchPart<mz_repr::Timestamp>>
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ptr/mod.rs:487:1
   4: core::ptr::drop_in_place<[mz_persist_client::fetch::LeasedBatchPart<mz_repr::Timestamp>]>
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ptr/mod.rs:487:1
   5: <alloc::vec::Vec<T,A> as core::ops::drop::Drop>::drop
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/alloc/src/vec/mod.rs:2920:13
   6: core::ptr::drop_in_place<alloc::vec::Vec<mz_persist_client::fetch::LeasedBatchPart<mz_repr::Timestamp>>>
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ptr/mod.rs:487:1
   7: core::ptr::drop_in_place<mz_persist_client::read::ReadHandle<mz_storage::types::sources::SourceData,(),mz_repr::Timestamp,i64>::subscribe::{{closure}}::{{closure}}>
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ptr/mod.rs:593:5
   8: core::ptr::drop_in_place<core::future::from_generator::GenFuture<mz_persist_client::read::ReadHandle<mz_storage::types::sources::SourceData,(),mz_repr::Timestamp,i64>::subscribe::{{closure}}::{{closure}}>>
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ptr/mod.rs:487:1
   9: core::ptr::drop_in_place<mz_persist_client::read::ReadHandle<mz_storage::types::sources::SourceData,(),mz_repr::Timestamp,i64>::subscribe::{{closure}}>
             at ./src/persist-client/src/read.rs:585:5
  10: core::ptr::drop_in_place<core::future::from_generator::GenFuture<mz_persist_client::read::ReadHandle<mz_storage::types::sources::SourceData,(),mz_repr::Timestamp,i64>::subscribe::{{closure}}>>
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ptr/mod.rs:487:1
  11: mz_storage::source::persist_source::persist_source_core::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/async-stream-0.3.3/src/lib.rs:237:9
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
  13: <async_stream::async_stream::AsyncStream<T,U> as futures_core::stream::Stream>::poll_next
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/async-stream-0.3.3/src/async_stream.rs:53:17
  14: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.25/src/stream.rs:120:9
  15: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.25/src/stream/stream/mod.rs:1626:9
  16: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.25/src/stream/stream/next.rs:32:9
  17: <mz_storage::source::persist_source::DropSafeLeaseStream as core::ops::drop::Drop>::drop::{{closure}}
             at ./src/storage/src/source/persist_source.rs:147:55
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
  19: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:165:17
  20: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/loom/std/unsafe_cell.rs:14:9
  21: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:155:13
  22: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:480:19
  23: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panic/unwind_safe.rs:271:9
  24: std::panicking::try::do_call
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:492:40
  25: std::panicking::try
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:456:19
  26: std::panic::catch_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panic.rs:137:14
  27: tokio::runtime::task::harness::poll_future
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:468:18
  28: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:104:27
  29: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:57:15
  30: tokio::runtime::task::raw::RawTask::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/raw.rs:84:18
  31: tokio::runtime::task::LocalNotified<S>::run
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/mod.rs:381:9
  32: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:435:13
  33: tokio::coop::with_budget::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/coop.rs:102:9
  34: std::thread::local::LocalKey<T>::try_with
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:445:16
  35: std::thread::local::LocalKey<T>::with
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:421:9
  36: tokio::coop::with_budget
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/coop.rs:95:5
  37: tokio::coop::budget
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/coop.rs:72:5
  38: tokio::runtime::thread_pool::worker::Context::run_task
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:434:9
  39: tokio::runtime::thread_pool::worker::Context::run
  40: tokio::runtime::thread_pool::worker::run::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:386:17
  41: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/macros/scoped_tls.rs:61:9
  42: tokio::runtime::thread_pool::worker::run
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:383:5
  43: tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:362:45
  44: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/task.rs:42:21
  45: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:165:17
  46: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/loom/std/unsafe_cell.rs:14:9
  47: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:155:13
  48: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:480:19
  49: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panic/unwind_safe.rs:271:9
  50: std::panicking::try::do_call
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:492:40
  51: std::panicking::try
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:456:19
  52: std::panic::catch_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panic.rs:137:14
  53: tokio::runtime::task::harness::poll_future
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:468:18
  54: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:104:27
  55: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:57:15
  56: tokio::runtime::task::raw::RawTask::poll
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/raw.rs:84:18
  57: tokio::runtime::task::UnownedTask<S>::run
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/mod.rs:418:9
  58: tokio::runtime::blocking::pool::Task::run
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/pool.rs:91:9
  59: tokio::runtime::blocking::pool::Inner::run
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/pool.rs:325:17
  60: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/pool.rs:300:13
```

**Comment by lluki** [comment]
`mz_persist_client::read::ReadHandle::subscribe` at `src/persist-client/src/read.rs:586` is an example where we hold the LeasedBatchPart across an await point, so that's a potential for the panic

**Comment by lluki** [comment]
To repro, build in release mode, then run this for at least 1000 iterations:
```
#!/usr/bin/python3
import psycopg2

PSQL="postgres://materialize@localhost:6875/materialize"
def sqls(conn, cmds):
    with conn.cursor() as cur:
        for cmd in cmds:
            cur.execute(cmd)

conn = psycopg2.connect(PSQL)
conn.autocommit = True
while True:
    sqls(conn, [
        "CREATE TABLE IF NOT EXISTS t1 (f1 int)",
        "CREATE MATERIALIZED VIEW IF NOT EXISTS v1 AS SELECT MIN(f1) FROM t1",
        "DROP MATERIALIZED VIEW v1",
        "DROP TABLE t1"
        ])
```

File: src/storage/src/source/persist_source.rs [hunk 1]
```diff
@@ -285,14 +290,25 @@ where
         // `as_of`.
         yield (Vec::new(), as_of_stream.clone());
 
-        let subscription = read.subscribe(as_of_stream.clone()).await;
+        // Always poll drop_rx first. drop_rx returns only if drop_tx has been dropped.
+        // In this case, we deliberately do not continue on the subscribe. but terminate this task
```
Comment (teskje) [inline src/storage/src/source/persist_source.rs:294]:
```suggestion
        // In this case, we deliberately do not continue on the subscribe, but terminate this task.
```

**Review by teskje — APPROVED** [review]

**Review by petrosagg — APPROVED** [review]
Looks good but we should only merge after this https://github.com/MaterializeInc/materialize/pull/15868 is merged

**Pull request merged**